src/linear_algebra/matrix/charpoly/linear_map.lean

```diff
@@ -0,0 +1,245 @@
+/-
+Copyright (c) 2022 Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Andrew Yang
+-/
+import linear_algebra.matrix.charpoly.coeff
+import linear_algebra.matrix.to_lin
+
+/-!
+
+# Calyley-Hamilton theorem for f.g. modules.
+
+Given a fixed finite spanning set `v : ι → M` of a `R`-module `M`, we say that a matrix `M`
+represents an endomorphism `f : M →ₗ[R] M` if the matrix as an endomorphism of `ι → R` commutes
+with `f` via the projection `(ι → R) →ₗ[R] M` given by `v`.
+
+We show that every endomorphism has a matrix representation, and if `f.range ≤ I • ⊤` for some
+ideal `I`, we may furthermore obtain a matrix representation whose entries fall in `I`.
+
+This is used to conclude the Cayley-Hamilton theorem for f.g. modules over arbitrary rings.
+-/
+
+variables {ι : Type*} [fintype ι]
+variables {M : Type*} [add_comm_group M] (R : Type*) [comm_ring R] [module R M] (I : ideal R)
+variables (v : ι → M) (hv : submodule.span R (set.range v) = ⊤)
+
+open_locale big_operators
+
+/-- Matrices, being endomorphisms of `ι → R`, acts on `(ι → R) →ₗ[R] M`, and takes the projection
+to a `(ι → R) →ₗ[R] M`.  -/
+def pi_to_module.from_matrix [decidable_eq ι] : matrix ι ι R →ₗ[R] (ι → R) →ₗ[R] M :=
+(linear_map.llcomp R _ _ _ (fintype.total R R v)).comp alg_equiv_matrix'.symm.to_linear_map
+
+lemma pi_to_module.from_matrix_apply [decidable_eq ι] (A : matrix ι ι R) (w : ι → R) :
+  pi_to_module.from_matrix R v A w = fintype.total R R v (A.mul_vec w) := rfl
+
+lemma pi_to_module.from_matrix_apply_single_one [decidable_eq ι] (A : matrix ι ι R) (i : ι) :
+  pi_to_module.from_matrix R v A (pi.single i 1) = ∑ (x : ι), A x i • v x :=
```

Comment:
And here I would think that `A i j` is a natural way to denote matrix entries. Could you please relabel the variable names?